### DISCUSSION TITLE
[A-L] (27 marzo 2024) Esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#27] [Esercitazioni A-L]
Sia un grafo **diretto** $G = (V, E)$ ed $s$ e $t$ due vertici di $G$. $G$ si dice $s-t$-connesso se ogni vertice di $G$ è in almeno un cammino da $s$ a $t$. Fornire un algoritmo in pseudo-codice che dato un grafo diretto e aciclico $G = (V, E)$ ed $s$ e $t$ due vertici di $G$, restituisce un sottografo $G'$ di $G$ massimale $s-t$-connesso

**Comment by LeoCosta02** [comment]
```C

//Algoritmo che seleziona il sottografo massimale che contiene
//tutti i vertici per cui passa almeno un cammino tra due determinati vertici.
Input: grafo diretto e aciclico G = (V, E) e due vertici s e t
Output: tutti i vertici che sono coinvolti in cammini da s a t
VERTICI_COINVOLTI(G,s,t){
    C[n]: array di n elementi inizializzato a -1
    L : list contenente i vertci coinvolti //Variabile globale non la passo alla funzione 
    C[t]=1
    if(s==t)// se s == t return subito uno dei due vertici 
    return s
    SELEZIONA_VERTICI(s,t)
    if (L==0){ //caso in cui non ho un cammino tra s e t metto nella lista solo i vertici s e t 
      L.add(s)
      L.add(t)
    }
     else{
      L.add(t)
     }
    return L 
}

SELEZIONA_VERTICI(u,t){
    if(C[u]!= -1){
        return C[u]
    }
    k=0
      for each v appartenente Adj[u]{
        k=k+CONTA_CAMMINI(v,t)

      }
      C[u]=k
      if(C[u]!=0){
        L.add(u)
      }
       return k
}
```

**Comment by Frisbon** [comment]
Parto da $s$ ed esploro con la DFS ricorsiva ogni singolo nodo, salvandomi in un insieme temporaneo i nodi visitati.
Se l'insieme temporaneo finisce con $t$, allora ho un cammino in cui i nodi appartengono al sottografo $s-t$ connesso. Altrimenti ritorno il vuoto.

Dopo aver trovato **tutti** i nodi del sottografo, basta prendere ogni singolo arco in cui questi compaiono.

![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/2d47b355-b1a5-4497-860e-bcde9eb884b1)

(non ho usato VIS nella ricorsione perchè ho bisogno di sapere se un nodo può allungare un percorso già esistente e anche perchè ho un DAG, impossibile che cicli)

**Comment by Elia-Belli** [comment]
## Programmazione Dinamica
$$Vis[k] = \text{appartenza del nodo k ad almeno un cammino da $s\to t$}$$

L'idea è di visitare $G$ a partire da $s$ incontrando tutti nodi raggiungibili da $s$, se da tali nodi è possibile raggiungere $t$ allora essi fanno parte del sottografo s-t-connesso, per verificare ciò si prosegue ricorsivamente fino a raggiungere $t$ (se possibile) e al ritorno si segna $Vis[k] = 1$ se è stato raggiunto $t$ nel suo sottoalbero di chiamate ricorsive, altrimenti $V[k]=0$.
```python fold
def conta_cammini(G,s,t):
    Vis = [-1] * n
    visita(G,Vis,s,t)

    if(Vis[s] = 0): return "Nessun sottografo s-t-connesso"

    Sol = {}
    for v in Vis:
        if(Vis[v] = 1):
            Sol.add(v)
    return Sol      #dati i nodi del sottografo, per gli archi è sufficiente recuperare
                    #gli archi in E(G) con entrambi i vertici in V(G')

	

def visita(G,Vis,k,t):
    if(Vis[k] != -1):
	return
		
    if (k = t):
	Vis[t] = 1
	return
	
    n = 0
    for v in k.adiacenti:
	visita(G,Vis,v,t)
	Vis[k] = max(n,Vis[v])
		
    return 
```
